File: java/Bit_Manipulation/Day3/Single_Number_III.java
package Bit_Manipulation.Day3;
import java.util.*;

public class Single_Number_III {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter the size of arr: ");
        int n = sc.nextInt();

        int[] arr = new int[n];
        System.out.print("Enter " + n + " elements: ");
        for (int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();
        }


        int[] res = single(arr);
        System.out.print("Single numbers are: " + Arrays.toString(res));
    }


    public static int[] single(int[] arr) {
        int n = arr.length;
        HashMap<Integer, Integer> mpp = new HashMap<>();

        for (int i = 0; i < n; i++) {
            mpp.put(arr[i], mpp.getOrDefault(arr[i], 0) + 1);
        }

        ArrayList<Integer> ans = new ArrayList<>();

        for (Map.Entry<Integer, Integer> it : mpp.entrySet()) {
            if (it.getValue() == 1) {
                ans.add(it.getKey());
            }
        }


        return new int[]{ans.get(0), ans.get(1)};
    }
}

File: python/Bit_Manipulation/day3/single_number_III.py
def single(arr):



    # brute force approach...
    # mpp = {}

    # for num in arr:
    #     mpp[num] = mpp.get(num, 0) + 1

    # ans = []

    # for key, value in mpp.items():
    #     if value == 1:
    #         ans.append(key)

    # return ans




    # optimal approach...
    xor = 0

    for num in arr:
        xor ^= num

    rightMost = xor & -xor

    x = 0
    y = 0

    for num in arr:
        if (num & rightMost) != 0:
            x ^= num
        else:
            y ^= num

    return [x, y]



# main
n = int(input("Enter the size of arr: "))
arr = list(map(int, input("Enter elements: ").split()))

res = single(arr)

print("Single numbers are:", res)